va/va.go

```diff
@@ -673,49 +685,157 @@ func (va *ValidationAuthorityImpl) performRemoteValidation(ctx context.Context,
+		if prob == nil {
 			good++
 		} else {
 			bad++
 		}
-		if good >= required {
-			result <- nil
-			state = "success"
-			break
-		} else if bad > va.maxRemoteFailures {
-			if prob, ok := err.(*probs.ProblemDetails); ok {
-				// The overall error returned is whichever error
-				// happened to tip the threshold. This is fine
-				// since we expect that any remote validation
-				// failures will typically be the same across
-				// instances.
-				result <- prob
-			} else {
-				// Otherwise the error was not an expected non-sucess problem result and
-				// represents an internal error. The real error has already been logged
-				// so return a server internal problem result without detail.
-				result <- probs.ServerInternal("Remote PerformValidation RPCs failed")
+
+		// Store the first non-nil problem to return later (if `MultiVAFullResults`
+		// is enabled).
+		if firstProb == nil && prob != nil {
+			firstProb = prob
+		}
+
+		// If MultiVAFullResults isn't enabled then return early whenever the
+		// success or failure threshold is met.
+		if !features.Enabled(features.MultiVAFullResults) {
+			if good >= required {
+				state = "success"
+				return nil
+			} else if bad > va.maxRemoteFailures {
+				return prob
 			}
+		}
+
+		// If we haven't returned early because of MultiVAFullResults being enabled
+		// we need to break the loop once all of the VAs have returned a result.
+		if len(remoteProbs) == len(va.remoteVAs) {
+			break
+		}
+	}
+
+	// If we are using `features.MultiVAFullResults` then we haven't returned
+	// early and can now log the differential between what the primary VA saw and
+	// what all of the remote VAs saw.
+	va.logRemoteValidationDifferentials(domain, primaryResult, remoteProbs)
+
+	// Based on the threshold of good/bad return nil or a problem.
+	if good >= required {
+		state = "success"
+		return nil
+	} else if bad > va.maxRemoteFailures {
+		return firstProb
+	}
+
+	// This condition should not occur - it indicates the good/bad counts didn't
+	// meet either the required threshold or the maxRemoteFailures threshold.
+	return probs.ServerInternal("Too few remote PerformValidation RPC results")
+}
+
+// logRemoteValidationDifferentials is called by `processRemoteResults` when the
+// `MultiVAFullResults` feature flag is enabled. It produces a JSON log line
+// that contains the primary VA result and the results each remote VA returned.
+func (va *ValidationAuthorityImpl) logRemoteValidationDifferentials(
+	domain string,
+	primaryResult *probs.ProblemDetails,
+	remoteProbs []*probs.ProblemDetails) {
+
+	allEqual := true
+	for _, e := range remoteProbs {
+		if e != primaryResult {
+			allEqual = false
 			break
 		}
 	}
+	if allEqual {
+		// There's no point logging a differential line if the primary VA and remote
+		// VAs all agree.
+		return
+	}
+
+	logOb := struct {
+		Domain        string
+		PrimaryResult *probs.ProblemDetails
+		RemoteResults []*probs.ProblemDetails
```

Comment:
We could remove some of the post-processing required to interpret this data by splitting `RemoteResults` into two fields, `RemoteSuccesses` and `RemoteErrors` (which would just be `RemoteResults` with the `nil` results filtered out). Not super important but I think it would make the investigatory phase slightly easier.

Reply:
Sounds good. Implemented in 24eedee